### DISCUSSION TITLE
KLU & BTF & OpenMP for the individual blocks

### DISCUSSION THREAD

**Discussion by DimaLPorNLP** [DrTimothyAldenDavis/SuiteSparse#884] [Ideas]
How many on average upper triangular blocks BTF generates in general? less than 10, less than 100, less than 1000? Could they be solved in parallel using OpenMP?

**Comment by DrTimothyAldenDavis** [comment]
Like all sparse matrix problems, it's impossible to say in general.  However, for most applications, there can be 1000's of blocks but it's usually the case that one of them is very large and the others tiny.  The advantage of BTF is that even when the blocks are tiny, there can be big win because the tiny diagonal blocks can correspond to very dense rows and or columns.  Say there are 2 blocks:

```
A = [ a11   a12
        0   A22 ]
```

and suppose a11 is a scalar.  Then if a12 is a dense row, this is a good thing because it will not appear in any of the factorizations.  It has been safely removed and cannot cause fillin.

It's rare, in practice, that the diagonal blocks are balanced in size, or that there are more than 1 substantial diagonal block.  So OpenMP is of little help to factorize the blocks in parallel.

It's always possible to make a contrived matrix that does have lots of big diagonal blocks, so OpenMP could be used and the algorithm would be sped up.  But these matrices rarely occur in practice, if at all.
